jepsen/src/jepsen/core.clj

```diff
@@ -138,72 +138,126 @@
-                    ; At this point all bets are off. If the client or network
-                    ; or DB crashed before doing anything; this operation won't
-                    ; be a part of the history. On the other hand, the DB may
-                    ; have applied this operation and we *don't know* about it;
-                    ; e.g.  because of timeout.
-                    ;
-                    ; This process is effectively hung; it can not initiate a
-                    ; new operation without violating the single-threaded
-                    ; process constraint. We cycle to a new process identifier,
-                    ; and leave the invocation uncompleted in the history.
-                    (conj-op! test (assoc op
-                                          :type :info
-                                          :time  (relative-time-nanos)
-                                          :error (str "indeterminate: "
-                                                      (if (.getCause t)
-                                                        (.. t getCause
-                                                            getMessage)
-                                                        (.getMessage t)))))
-                    (warn t "Process" process "indeterminate")
-                    (+ process (:concurrency test))))))))
-        (info "Worker" process "done")))))
+        (info "Worker" process "with node" node "starting")
+
+        ; Differentiate our base client for this process
+        (let [client (client/open-compat! (:client test) test node)
+              ; Wait to begin ops until all clients are set up
+              _ (.await setup-barrier)
+
+              ; Consume operations from the generator
+              client (try
+                       (loop [[process client] [process client]]
+                         ; Obtain an operation to execute
+                         (if-let [op (generator/op-and-validate gen test process)]
+                           (let [op (assoc op
+                                           :process process
+                                           :time    (relative-time-nanos))]
+                             ; Log invocation
+                             (util/log-op op)
+                             ; Add invocation to history
+                             (conj-op! test op)
+                             ; Apply the op and log its completion in the history
+                             (recur (invoke-and-complete node process client test op)))
+
+                           ; Out of ops, return most recent client
+                           client))
+
+                       ; Exceptions here aren't expected, clean up state then re-throw
+                       (catch Exception e
+                         (warn "Worker for" process "threw unexpectedly. Tearing down")
+                         (.await setup-barrier)
+                         (client/close-compat! client test)
```

Comment:
So the client that we tear down here might be a *different* client from the one inside the loop/recur. What about... maybe having a third loop/recur argument, which is nil normally, and an exception if we throw? Then we could do

```clj
(let [[process client exception]
  (loop [[process client exception] [process client nil]]
     (try ... (catch Exception e [process client e]))))
```

Reply:
You're right that the wrong client might get closed out here... I'd need a result from the loop even though it threw dang.

When attempting to pass the exception up for later, I'm having a difficult time getting around 'cannot recur across try'. I'm having a difficult time thinking around this, any ideas?

Reply:
Ok, got something with (recur (try ...)) and returning a bool to signal our exit condition